### PR TITLE
Add service discovery

### PR DESCRIPTION
**Pull request by ezmac**
##### SUMMARY

Adds two new modules in the cloud/amazon namespace, service_discovery and service_discovery_namespace with the aim of allowing these to be created more easily.  This is in preparation to add service discovery support to ecs_service.

It might be possible to do both of these in one module, but two made more sense to me when I was working on it.

Not sure if this is written to how the community wants modules to work, but I'm open to feedback.

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
service_discovery
service_discovery_namespace

##### ANSIBLE VERSION
```paste below
ansible 2.8.0.dev0 (add_service_discovery 6e62b5a7b7) last updated 2018/11/07 15:50:32 (GMT -400)
  config file = None
  configured module search path = ['/home/tad/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/tad/code/ansible/lib/ansible
  executable location = /home/tad/code/ansible/bin/ansible
  python version = 3.5.2 (default, Nov 23 2017, 16:37:01) [GCC 5.4.0 20160609]

```

##### ADDITIONAL INFORMATION

https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/servicediscovery.html

### CODE REVIEW

**Comment by willthames** [comment]
I don't understand why the test suite is added to ecs_cluster rather than having its own test suite - as far as I can tell it's independent and not relying on any aspect of ecs_cluster.

File: lib/ansible/modules/cloud/amazon/service_discovery.py [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+
+
+from ansible.module_utils.aws.core import AnsibleAWSModule
+from ansible.module_utils.ec2 import ec2_argument_spec
+from ansible.module_utils.ec2 import snake_dict_to_camel_dict, camel_dict_to_snake_dict, map_complex_type, get_ec2_security_group_ids_from_names
+from ansible.module_utils.ec2 import ansible_dict_to_boto3_filter_list
+
+try:
+    import botocore
+    import boto3
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/service_discovery.py:222]: Unused, as far as I can tell

File: lib/ansible/modules/cloud/amazon/service_discovery.py [hunk 2]
```diff
@@ -0,0 +1,402 @@
...
+        if 'events' in service:
+            for e in service['events']:
+                if 'createdAt' in e:
+                    e['createdAt'] = str(e['createdAt'])
+        return service
+
+
+    # Boto provides these functions:
+
+    # can_paginate()
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/service_discovery.py:330]: Please remove these comments

File: lib/ansible/modules/cloud/amazon/service_discovery_namespace.py [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+        state=dict(required=True, choices=['present', 'absent']),
+        type=dict(required=False, type='str',choices=['DNS_PRIVATE', 'DNS_PUBLIC']),
+        id=dict(required=False, type=str),
+        name=dict(required=True, type='str'),
+        creator_request_id=dict(required=False, type='str'),
+        description=dict(required=False, type='str'),
+        wait=dict(required=False, type='bool', default=True),
+        vpc_id=dict(required=False, type='str')
+    ))
+    # namespaces sssuuuuucccckkkk.
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/service_discovery_namespace.py:305]: Please remove this comment

File: lib/ansible/modules/cloud/amazon/service_discovery_namespace.py [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+        return service
+
+
+
+def main():
+    argument_spec = ec2_argument_spec()
+    argument_spec.update(dict(
+        state=dict(required=True, choices=['present', 'absent']),
+        type=dict(required=False, type='str',choices=['DNS_PRIVATE', 'DNS_PUBLIC']),
+        id=dict(required=False, type=str),
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/service_discovery_namespace.py:298]: required=False, type='str' are defaults, and not needed in argument_specs

File: lib/ansible/modules/cloud/amazon/service_discovery_namespace.py [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+            waiter.wait(OperationId=operation)
+            operationResponse = self.sd.get_operation(OperationId=operation)
+            namespace_id = operationResponse['Operation']['Targets']["NAMESPACE"]
+            created_namespace = self.sd.get_namespace(Id = namespace_id)['Namespace']
+            return created_namespace
+        else:
+            return response
+        
+    def create_private_dns_namespace(self, name, description, creator_request_id, vpc_id, wait):
+        # creator_req_id totes not needed, will be auto poulated if not given
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/service_discovery_namespace.py:242]: typo: populated

File: lib/ansible/modules/cloud/amazon/service_discovery_namespace.py [hunk 6]
```diff
@@ -0,0 +1,348 @@
...
+    def list_namespaces(self, type):
+
+        if type:
+            filters = ansible_dict_to_boto3_filter_list({'TYPE': type})
+        else:
+            filters = []
+        return self.sd.list_namespaces(Filters=filters)
+
+    def create_public_dns_namespace(self, name, description, creator_request_id, wait):
+        # creator_req_id totes not needed, will be auto poulated if not given
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/service_discovery_namespace.py:216]: typo: populated

**Review by willthames — CHANGES_REQUESTED** [review]

**Comment by willthames** [comment]
I should have spotted the failing tests before I started my review (normally reviewers aren't notified before tests pass).

This PR needs splitting into two and the various test failures addressing.

**Comment by ezmac** [comment]
Hey @willthames,

Thanks for the feedback.  I'm willing to put in the work to make this two PRs, but with service_discovery needing a namespace for any kind of testing, how do you suggest I write a test suite?  I can use shell/command module and create one with awscli, but that sounds like a bad idea.  Does it make sense to send the first PR/branch for service_discovery_namespace, branch from that for service_discovery and submit a second PR that continues the first?  Any other approach?

I did the tests in ecs_cluster because I knew it mostly worked and it would save me time.  I may have missed documentation, but there's a lot of "figure it out" overhead getting started.  I'm trying to keep notes, so if you have a place I can publish them, I'll do what I can.

**Pull request closed without merging**